# Milestone 3: Exploratory Data Analysis — Orders

This notebook performs a limited, beginner-friendly inspection of only `olist_orders_dataset.csv`. It does not clean, modify, visualize, or engineer features from the data.

## 1. Import Libraries

**Why:** pandas provides the table-reading and inspection tools needed for this notebook.

**What we expect to learn:** This step only prepares the notebook; it does not inspect or change any business data.

In [1]:
import pandas as pd

**Meaning:** pandas is available and ready to load the orders table. No data has been read or changed yet.

## 2. Load Dataset

**Why:** We need the orders CSV in a DataFrame so we can inspect its structure.

**What we expect to learn:** This creates an in-memory view of the orders table. It reads only the approved CSV and does not write back to it.

In [2]:
orders_path = "../data/raw/olist/olist_orders_dataset.csv"
orders_df = pd.read_csv(orders_path)

**Meaning:** The raw orders table is loaded in memory for read-only inspection. No cleaning or modification has been applied.

## 3. Display Shape

**Why:** Shape gives the number of order records and fields at a glance.

**What we expect to learn:** We expect a `(rows, columns)` pair describing the size of this table.

In [3]:
orders_df.shape

(99441, 8)

**Business meaning:** The table contains 99,441 order records described by 8 fields. This establishes the scale of the order-lifecycle data available for later analysis.

## 4. Display First 5 Rows

**Why:** A small preview helps confirm how one order record is organized without analyzing the full dataset.

**What we expect to learn:** We expect to see identifiers, status, and lifecycle timestamps in their raw form.

In [4]:
orders_df.head()

                           order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   
  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   1          2018-07-26 14:31:00           2018-08-07 15:27:4

**Business meaning:** Each row represents an order linked to a customer, a current/final status, and timestamps for major fulfilment stages. The preview confirms that the table follows the order journey from purchase to estimated and actual delivery.

## 5. Display Column Names

**Why:** Column names define the business fields available in the orders table.

**What we expect to learn:** We expect identifiers, order status, and timestamps covering the order lifecycle.

In [5]:
orders_df.columns

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date'],
      dtype='str')

**Business meaning:** The available fields support order identification, customer linkage, status monitoring, and measurement of approval and delivery stages. No sales value or product detail exists in this table.

## 6. Display Data Types

**Why:** Data types show how pandas interpreted each raw column.

**What we expect to learn:** We expect identifiers, status, and unparsed timestamp fields to appear as text because no conversion is allowed in this milestone.

In [6]:
orders_df.dtypes

order_id                         str
customer_id                      str
order_status                     str
order_purchase_timestamp         str
order_approved_at                str
order_delivered_carrier_date     str
order_delivered_customer_date    str
order_estimated_delivery_date    str
dtype: object

**Business meaning:** All eight fields are currently text. The lifecycle dates cannot yet support time calculations without a later, explicitly approved conversion step; this notebook intentionally leaves them unchanged.

## 7. Check Missing Values

**Why:** Missing-value counts reveal which order-lifecycle fields are not populated for every record.

**What we expect to learn:** We expect identifiers and purchase fields to be complete, while some later lifecycle timestamps may be absent for orders that did not reach those stages.

In [7]:
orders_df.isna().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

**Business meaning:** Order IDs, customer IDs, statuses, purchases, and estimates are complete. Approval is missing for 160 records, carrier handoff for 1,783, and customer delivery for 2,965. These gaps may reflect orders that were canceled, unavailable, or still in progress, but that explanation must be tested later rather than assumed.

## 8. Count Unique Values for `order_status`

**Why:** Status counts show how orders are distributed across fulfilment stages and outcomes.

**What we expect to learn:** We expect to identify the available status categories and the number of records in each category.

In [8]:
orders_df["order_status"].value_counts(dropna=False)

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

**Business meaning:** The table contains eight statuses and is dominated by delivered orders (96,478). Smaller groups represent shipped, canceled, unavailable, invoiced, processing, created, and approved orders, which are important exceptions or unfinished stages for later investigation.

## 9. Basic Summary Using `describe()`

**Why:** `describe()` provides a consistent structural summary of count, distinct values, most common value, and frequency for the text columns.

**What we expect to learn:** We expect to confirm completeness, identifier uniqueness, status variety, and repeated timestamp values without changing data types.

In [9]:
orders_df.describe(include="all")

                                order_id                       customer_id  \
count                              99441                             99441   unique                             99441                             99441   top     e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   freq                                   1                                 1   
       order_status order_purchase_timestamp    order_approved_at  \
count         99441                    99441                99281   unique            8                    98875                90733   top       delivered      2018-03-31 15:08:21  2018-02-27 04:31:10   freq          96478                        3                    9   
       order_delivered_carrier_date order_delivered_customer_date  \
count                         97658                         96476   unique                        81018                         95664   top             2018-05-09 15:48:00           2018-05-14 20:02:44 

**Business meaning:** Both order and customer IDs are unique in this table, confirming one order-level customer record per order. Status has eight categories, and delivered is the most frequent. Lower counts in later lifecycle timestamps agree with the missing-value check. Repeated timestamps are possible and do not by themselves indicate duplicate orders.

## 10. Business Observations

- The orders table contains 99,441 unique orders and 99,441 unique order-level customer IDs.
- Most records have a `delivered` status, while a smaller set represents unfinished or unsuccessful order outcomes.
- Purchase and estimated-delivery timestamps are complete, but approval, carrier handoff, and actual-delivery timestamps contain missing values.
- Missing later-stage timestamps should be evaluated together with order status in the next approved EDA step; they should not be filled or removed without understanding the business process.
- All timestamp columns remain raw text. Converting them to datetime would be a separate preparation decision and was intentionally not performed here.
- This table describes order progress but does not contain product, seller, payment, review, customer-location, or monetary detail. No other table was analyzed.